assets/elm/src/Checkpoints.elm

```diff
@@ -0,0 +1,459 @@
+module Checkpoints exposing (main)
+
+-- Display route checkpoints
+--
+--
+
+import Browser
+import GeoJson exposing (GeoJson)
+import Html exposing (..)
+import Html.Attributes exposing (attribute, checked, class, classList, for, id, name, type_, value, width)
+import Html.Events exposing (onClick)
+import Http exposing (Error)
+import Json.Decode as Decode exposing (Decoder)
+import Json.Decode.Pipeline exposing (required)
+import Json.Encode exposing (Value)
+import Numeral exposing (format)
+import Set exposing (Set)
+
+
+
+-- MAIN
+
+
+main =
+    Browser.element
+        { init = init
+        , update = update
+        , subscriptions = \_ -> Sub.none
+        , view = view
+        }
+
+
+init : Value -> ( Model, Cmd Msg )
+init flags =
+    let
+        flags_result =
+            Decode.decodeValue configDecoder flags
+    in
+    case flags_result of
+        Ok config ->
+            let
+                url =
+                    config.displayUrl
+
+                cmd =
+                    getCheckpoints url GotExistingCheckpoints
+            in
+            ( Model config LoadingExistingCheckpoints, cmd )
+
+        Err error ->
+            let
+                config =
+                    Config "" "" "" False
+
+                errorString =
+                    Decode.errorToString error
+            in
+            ( Model config (Failure errorString), Cmd.none )
+
+
+
+-- MODEL
+
+
+type alias Model =
+    { config : Config
+    , status : Status
+    }
+
+
+type alias Config =
+    { displayUrl : String
+    , editUrl : String
+    , csrfToken : String
+    , canEdit : Bool
+    }
+
+
+configDecoder : Decoder Config
+configDecoder =
+    Decode.succeed Config
+        |> required "display_url" Decode.string
+        |> required "edit_url" Decode.string
+        |> required "csrf_token" Decode.string
+        |> required "can_edit" Decode.bool
+
+
+type Status
+    = Failure String
+    | LoadingExistingCheckpoints
+    | Display (List Checkpoint)
+    | LoadingPossibleCheckpoints (List Checkpoint)
+    | Edit CheckpointSelection
+    | Saving
+
+
+type alias CheckpointSelection =
+    { checkpoints : List Checkpoint, selected : Set FieldValue }
```

Comment:
C’est du pinaillage, mais cette structure rend théoriquement possible d’avoir des checkpoints dans `selected` sans qu’ils soient dans `checkpoints`. Pour éviter ça, tu pourrais utiliser une structure comme ça: `type alias CheckpointSelection = List ( Checkpoint, Bool )`

Reply:
C'est inspiré d'Ewan: https://youtu.be/XpDsk374LDE?t=1161

Et en fait, c’est un cas tout à fait possible théoriquement, quand tu mets à jour le parcours de ta route depuis le service distant. Mais impossible en pratique, parce que tu construits le set avec les checkpoints.

Je voulais attendre d'implémenter les checkpoints sur l'update de la route avant de tout changer. Je me la garde sous le coude…

Reply:
Ahhh fruits.com <3